src/services/product.services.new.test.ts

```diff
@@ -0,0 +1,74 @@
+import { faker } from '@faker-js/faker';
+import { db } from '../../prisma/db';
+import { ApiError } from '../config/apiError';
+import { errors } from '../config/errors';
+import { ProductSerializer } from '../serializers/product-serializer';
+import { productRandomRaw } from '../test/test-constants';
+import { UpdateProductRequest } from '../types/product';
+import { ProductService } from './product.services';
+
+describe('Products Service', () => {
+  describe('Update Product Service', () => {
+    const productId = faker.number.int();
+
+    const productRequest: UpdateProductRequest = {
+      title: 'newTitle',
+      description: ' newDescription',
+    };
+
+    const productEditedRaw = {
+      id: productId,
+      userId: faker.number.int(),
+      title: productRequest.title,
+      description: productRequest.description,
+    };
+
+    describe('Successful case', () => {
+      it('should call findUnique with correct arguments', async () => {
+        const findUniqueSpy = jest.spyOn(db.product, 'findUnique').mockResolvedValue(productRandomRaw);
+        jest.spyOn(db.product, 'update').mockResolvedValue(productEditedRaw);
+
+        await ProductService.updateProductService(productId, productRequest);
+
+        expect(findUniqueSpy).toHaveBeenCalledWith({ where: { id: productId } });
+      });
+
+      it('should call update with correct arguments', async () => {
+        jest.spyOn(db.product, 'findUnique').mockResolvedValue(productRandomRaw);
+        const updateSpy = jest.spyOn(db.product, 'update').mockResolvedValue(productEditedRaw);
+
+        await ProductService.updateProductService(productId, productRequest);
+
+        expect(updateSpy).toHaveBeenCalledWith({ data: { ...productRequest }, where: { id: productId } });
+      });
+
+      it('should return the edited product', async () => {
+        jest.spyOn(db.product, 'findUnique').mockResolvedValue(productRandomRaw);
+        jest.spyOn(db.product, 'update').mockResolvedValue(productEditedRaw);
+
+        const result = await ProductService.updateProductService(productId, productRequest);
+
+        expect(result).toEqual(ProductSerializer.serialize(productEditedRaw));
+      });
+    });
+
+    describe('Error case', () => {
+      it('should throw an ApiError (NOT_FOUND) when product is not found', async () => {
```

Comment:
When the product is not found, the update method should not be called (here I have a question because I don't recall how it behaves, if the update is with an id that doesn't exist, does Prisma return an error or an empty result?)

Reply:
![imagen](https://github.com/miranda-daniel/node-training-daniel-m/assets/11562287/189bb50a-3a16-46e1-bc66-1dab91f3a902)
This is the entire updateProductService method. When the productId is not found in the database by findUnique, Prisma returns null. In that case, throw new ApiError(errors.NOT_FOUND); is executed, and the exception is propagated back to the controller. Consequently, the update operation is never executed. 
That's why in the unit test, I don't even mock the Prisma update method; it's because that part of the code is never reached.